BUG fixes tuple agg  issue 18079

bobhaffner opened the pull request:
- [x] closes #18079
- [x] tests added / passed
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

TomAugspurger commented on doc/source/whatsnew/v0.21.1.txt:
```diff
@@ -62,6 +62,7 @@ Bug Fixes
 - Bug in ``pd.Series.rolling.skew()`` and ``rolling.kurt()`` with all equal values has floating issue (:issue:`18044`)
 - Bug in ``pd.DataFrameGroupBy.count()`` when counting over a datetimelike column (:issue:`13393`)
 - Bug in ``pd.concat`` when empty and non-empty DataFrames or Series are concatenated (:issue:`18178` :issue:`18187`)
+- Bug in :class:`NDFrameGroupBy` fixes ValueError: no results error when grouping by a single column and aggregating with a tuple (:issue:`18079`)  
```
Maybe just say something like "Bug when grouping by a single column and aggregating with a class like ``list`` or ``tuple``" . I don't recall if NDFRameGroupBy is in the public API.

TomAugspurger commented on pandas/tests/groupby/test_groupby.py:
```diff
@@ -2718,6 +2718,22 @@ def test_empty_dataframe_groupby(self):
+        assert_frame_equal(result, expected)
+
+        result = df.groupby('A')['C'].aggregate(tuple)
+        expected = pd.Series([(1, 1, 1), (3, 4, 4)], index=[1, 3])
```
Should be able to pass `name='C'` here.

bobhaffner commented:
@TomAugspurger's  comment on whats new prompted me to add a test for `list` aggregation only to discover a new issue.   

This works
df.groupby('A')['C'].aggregate(list)

This doesn't.  
df.groupby(['A', 'B']).aggregate(list)

Should I submit a separate issue for this or investigate and fix this new error as part of this PR?

<details>
  <summary>Click to see full traceback</summary>
  ---------------------------------------------------------------------------

Traceback (most recent call last):
  File "/Users/bob/dev/pandas-dev/pandas/core/groupby.py", line 2267, in agg_series
    return self._aggregate_series_fast(obj, func)
  File "/Users/bob/dev/pandas-dev/pandas/core/groupby.py", line 2287, in _aggregate_series_fast
    result, counts = grouper.get_result()
  File "pandas/_libs/src/reduce.pyx", line 406, in pandas._libs.lib.SeriesGrouper.get_result
    raise
  File "pandas/_libs/src/reduce.pyx", line 394, in pandas._libs.lib.SeriesGrouper.get_result
    result = _get_result_array(res,
  File "pandas/_libs/src/reduce.pyx", line 15, in pandas._libs.lib._get_result_array
    raise ValueError('function does not reduce')
ValueError: function does not reduce

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/bob/dev/pandas-dev/pandas/core/groupby.py", line 4192, in aggregate
    return super(DataFrameGroupBy, self).aggregate(arg, *args, **kwargs)
  File "/Users/bob/dev/pandas-dev/pandas/core/groupby.py", line 3634, in aggregate
    return self._python_agg_general(arg, *args, **kwargs)
  File "/Users/bob/dev/pandas-dev/pandas/core/groupby.py", line 853, in _python_agg_general
    result, counts = self.grouper.agg_series(obj, f)
  File "/Users/bob/dev/pandas-dev/pandas/core/groupby.py", line 2269, in agg_series
    return self._aggregate_series_pure_python(obj, func)
  File "/Users/bob/dev/pandas-dev/pandas/core/groupby.py", line 2304, in _aggregate_series_pure_python
    raise ValueError('Function does not reduce')
ValueError: Function does not reduce

</details>

jreback commented on pandas/tests/groupby/test_groupby.py:
```diff
@@ -2718,6 +2718,21 @@ def test_empty_dataframe_groupby(self):
+        # Issue #18079
+        df = pd.DataFrame({'A': [1, 1, 1, 3, 3, 3],
+                           'B': [1, 1, 1, 4, 4, 4], 'C': [1, 1, 1, 3, 4, 4]})
+
```
move to test_aggregate

can you parametrize with list-likes here (e.g. tuple/list,np.array,Series)

some of these might have different output (Series) but let's see.

you may have to have a tests with the parameterized ones and another one with 'other' things, e.g. 

```
def f(x):
    return tuple(x)
```

which work now.

bobhaffner replied:
Thanks for the feedback, @jreback!  I'm hoping to tackle the list bug and these additional tests later this week.

bobhaffner committed:
fix list agg and new tests

bobhaffner committed:
added whatsnew

bobhaffner commented:
- Moved tests to test_aggregate.py

- Parametrize the tests (one for DataFrames and one for Series))

- I fixed the list issue I mentioned above by removing the list condition in the following if

https://github.com/pandas-dev/pandas/blob/4fce7846be56e12999fe8758abb2ea2f2794259d/pandas/core/groupby.py#L2303

I was a little leery doing this, but all tests passed.

Note: Both np.array and pd.Series still error out with both scenarios (grouping by one column vs multiple columns), but with different error descriptions.  That said, is it possible to proceed with this PR since both tuple and list now behave the same way with single and multiple column groupbys?
```
df = pd.DataFrame({'A' : [1, 1, 3], 'B' :  [1, 1, 4], 'C' :  [1, 3, 4]})

df.groupby('A')['C'].aggregate(pd.Series) # np.array
>> Exception: Must produce aggregated value

df.groupby(['A', 'B']).aggregate(pd.Series) # np.array
>> ValueError: Function does not reduce
```

This might close https://github.com/pandas-dev/pandas/issues/4293 as well

jreback commented on doc/source/whatsnew/v0.21.1.txt:
```diff
@@ -103,7 +103,7 @@ Groupby/Resample/Rolling
 - Bug in ``DataFrame.resample(...)`` when there is a time change (DST) and resampling frequecy is 12h or higher (:issue:`15549`)
 - Bug in ``pd.DataFrameGroupBy.count()`` when counting over a datetimelike column (:issue:`13393`)
--
+- Bug when grouping by a single column and aggregating with a class like`list` or `tuple` (:issue:`18079`)
```
use double-backticks around ``list`` and ``tuple``

jreback replied:
move this to 0.22 as well. (its not that I don't regard this as a bug fix, but this is something touching an important piece of code and want this to live in master for a bit).

jreback reviewed (changes requested):
small changes, otherwise lgtm. ping on green.

bobhaffner committed:
backticks and moved to .22

bobhaffner commented:
Thanks @jreback!

jreback commented:
one more rebase :>

bobhaffner committed:
rebase cleanup

pep8speaks commented:
Hello @bobhaffner! Thanks for updating the PR.





 - In the file [`test_agg.py`](https://github.com/pandas-dev/pandas/blob/6639d48801877fb85f8bbe7052bbb07a435cfdac/test_agg.py), following are the PEP8 issues :

> [Line 4:1](https://github.com/pandas-dev/pandas/blob/6639d48801877fb85f8bbe7052bbb07a435cfdac/test_agg.py#L4): [E302](https://duckduckgo.com/?q=pep8%20E302) expected 2 blank lines, found 1
> [Line 5:4](https://github.com/pandas-dev/pandas/blob/6639d48801877fb85f8bbe7052bbb07a435cfdac/test_agg.py#L5): [E111](https://duckduckgo.com/?q=pep8%20E111) indentation is not a multiple of four
> [Line 7:1](https://github.com/pandas-dev/pandas/blob/6639d48801877fb85f8bbe7052bbb07a435cfdac/test_agg.py#L7): [E265](https://duckduckgo.com/?q=pep8%20E265) block comment should start with '# '
> [Line 8:1](https://github.com/pandas-dev/pandas/blob/6639d48801877fb85f8bbe7052bbb07a435cfdac/test_agg.py#L8): [E265](https://duckduckgo.com/?q=pep8%20E265) block comment should start with '# '
> [Line 11:1](https://github.com/pandas-dev/pandas/blob/6639d48801877fb85f8bbe7052bbb07a435cfdac/test_agg.py#L11): [E265](https://duckduckgo.com/?q=pep8%20E265) block comment should start with '# '
> [Line 12:1](https://github.com/pandas-dev/pandas/blob/6639d48801877fb85f8bbe7052bbb07a435cfdac/test_agg.py#L12): [E265](https://duckduckgo.com/?q=pep8%20E265) block comment should start with '# '
> [Line 13:1](https://github.com/pandas-dev/pandas/blob/6639d48801877fb85f8bbe7052bbb07a435cfdac/test_agg.py#L13): [E265](https://duckduckgo.com/?q=pep8%20E265) block comment should start with '# '
> [Line 15:23](https://github.com/pandas-dev/pandas/blob/6639d48801877fb85f8bbe7052bbb07a435cfdac/test_agg.py#L15): [E203](https://duckduckgo.com/?q=pep8%20E203) whitespace before ':'
> [Line 15:40](https://github.com/pandas-dev/pandas/blob/6639d48801877fb85f8bbe7052bbb07a435cfdac/test_agg.py#L15): [E203](https://duckduckgo.com/?q=pep8%20E203) whitespace before ':'
> [Line 15:42](https://github.com/pandas-dev/pandas/blob/6639d48801877fb85f8bbe7052bbb07a435cfdac/test_agg.py#L15): [E241](https://duckduckgo.com/?q=pep8%20E241) multiple spaces after ':'
> [Line 15:58](https://github.com/pandas-dev/pandas/blob/6639d48801877fb85f8bbe7052bbb07a435cfdac/test_agg.py#L15): [E203](https://duckduckgo.com/?q=pep8%20E203) whitespace before ':'
> [Line 15:60](https://github.com/pandas-dev/pandas/blob/6639d48801877fb85f8bbe7052bbb07a435cfdac/test_agg.py#L15): [E241](https://duckduckgo.com/?q=pep8%20E241) multiple spaces after ':'
> [Line 16:1](https://github.com/pandas-dev/pandas/blob/6639d48801877fb85f8bbe7052bbb07a435cfdac/test_agg.py#L16): [E265](https://duckduckgo.com/?q=pep8%20E265) block comment should start with '# '
> [Line 19:1](https://github.com/pandas-dev/pandas/blob/6639d48801877fb85f8bbe7052bbb07a435cfdac/test_agg.py#L19): [E265](https://duckduckgo.com/?q=pep8%20E265) block comment should start with '# '
> [Line 20:2](https://github.com/pandas-dev/pandas/blob/6639d48801877fb85f8bbe7052bbb07a435cfdac/test_agg.py#L20): [E114](https://duckduckgo.com/?q=pep8%20E114) indentation is not a multiple of four (comment)
> [Line 20:2](https://github.com/pandas-dev/pandas/blob/6639d48801877fb85f8bbe7052bbb07a435cfdac/test_agg.py#L20): [E116](https://duckduckgo.com/?q=pep8%20E116) unexpected indentation (comment)
> [Line 22:1](https://github.com/pandas-dev/pandas/blob/6639d48801877fb85f8bbe7052bbb07a435cfdac/test_agg.py#L22): [E265](https://duckduckgo.com/?q=pep8%20E265) block comment should start with '# '
> [Line 24:1](https://github.com/pandas-dev/pandas/blob/6639d48801877fb85f8bbe7052bbb07a435cfdac/test_agg.py#L24): [E265](https://duckduckgo.com/?q=pep8%20E265) block comment should start with '# '
> [Line 27:1](https://github.com/pandas-dev/pandas/blob/6639d48801877fb85f8bbe7052bbb07a435cfdac/test_agg.py#L27): [E265](https://duckduckgo.com/?q=pep8%20E265) block comment should start with '# '
> [Line 32:1](https://github.com/pandas-dev/pandas/blob/6639d48801877fb85f8bbe7052bbb07a435cfdac/test_agg.py#L32): [E302](https://duckduckgo.com/?q=pep8%20E302) expected 2 blank lines, found 1
> [Line 33:4](https://github.com/pandas-dev/pandas/blob/6639d48801877fb85f8bbe7052bbb07a435cfdac/test_agg.py#L33): [E111](https://duckduckgo.com/?q=pep8%20E111) indentation is not a multiple of four
> [Line 35:1](https://github.com/pandas-dev/pandas/blob/6639d48801877fb85f8bbe7052bbb07a435cfdac/test_agg.py#L35): [E305](https://duckduckgo.com/?q=pep8%20E305) expected 2 blank lines after class or function definition, found 1
> [Line 35:6](https://github.com/pandas-dev/pandas/blob/6639d48801877fb85f8bbe7052bbb07a435cfdac/test_agg.py#L35): [E211](https://duckduckgo.com/?q=pep8%20E211) whitespace before '('
> [Line 37:6](https://github.com/pandas-dev/pandas/blob/6639d48801877fb85f8bbe7052bbb07a435cfdac/test_agg.py#L37): [E211](https://duckduckgo.com/?q=pep8%20E211) whitespace before '('
> [Line 44:1](https://github.com/pandas-dev/pandas/blob/6639d48801877fb85f8bbe7052bbb07a435cfdac/test_agg.py#L44): [W391](https://duckduckgo.com/?q=pep8%20W391) blank line at end of file

bobhaffner commented:
Hi @jreback I had conflicts (whatsnew) with the rebase, but I believe Its corrected now.  That said, I'm puzzled by the pep8speaks comment regarding `test_agg.py` though as I didn't modify that file.  My apologies for the confusion.

jreback commented on doc/source/whatsnew/v0.21.1.txt.orig:
```diff
@@ -0,0 +1,154 @@
+
+.. _whatsnew_0211.bug_fixes:
+
+Bug Fixes
```
remove this

jreback commented on grp.patch:
```diff
@@ -0,0 +1,13 @@
+         if isinstance(func_or_funcs, compat.string_types):
+             return getattr(self, func_or_funcs)(*args, **kwargs)
+ 
+-        if hasattr(func_or_funcs, '__iter__'):
```
remove this

jreback commented on test_agg.py:
```diff
@@ -0,0 +1,44 @@
+   return list(x)
+
+#df = pd.DataFrame({'A' : [1, 1, 3], 'B' :  [1, 2, 4]})
+#result = df.groupby('A').aggregate(f)
```
remove this

jreback reviewed (changes requested):
looks like some extra files got added during the rebase. pls remove them. and push again.

bobhaffner committed:
removed local files

bobhaffner commented:
Yikes, my mistake.  thanks Jeff.

jreback commented on grp_test.patch:
```diff
@@ -0,0 +1,15 @@
+@@ -2725,3 +2725,12 @@ def _check_groupby(df, result, keys, field, f=lambda x: x.sum()):
+     expected = f(df.groupby(tups)[field])
+     for k, v in compat.iteritems(expected):
+         assert (result[k] == v)
```
patch needs removal

jreback commented on doc/source/whatsnew/v0.21.1.txt:
```diff
@@ -147,4 +147,4 @@ Other
 ^^^^^
 
 -
```
can you revert this file

bobhaffner replied:
HI Jeff, I'm afraid my lack of git experience is showing as this file has been a bit of a thorn in my side.   I'm concerned that my attempts to fix my mess will just result in wasting more of your time.  Would you mind sharing the syntax necessary to revert (or reset?) this file to the proper commit?

jreback replied:
sure ill fix this upl

jreback reviewed (changes requested):
couple of more items to remove

jreback committed:
Merge branch 'master' into PR_TOOL_MERGE_PR_18354

jreback committed:
fixups

jreback commented:
should be good to go. ping on green.

bobhaffner commented:
Thanks Jeff!  You're a lifesaver!

jreback commented:
thanks @bobhaffner

bobhaffner commented:
Many thanks for the help and the patience, @jreback .   Same to you @TomAugspurger
